src/ucp/tag/rndv.c

```diff
@@ -221,27 +221,94 @@ static void ucp_rndv_req_send_rtr(ucp_request_t *rndv_req, ucp_request_t *rreq,
     ucp_request_send(rndv_req);
 }
 
+static void ucp_rndv_get_lanes_count(ucp_request_t *req)
+{
+    ucp_ep_h ep        = req->send.ep;
+    ucp_lane_map_t map = 0;
+    uct_rkey_t uct_rkey;
+    ucp_lane_index_t lane;
+
+    if (ucs_likely(req->send.rndv_get.lane_count != 0)) {
+        return; /* already resolved */
+    }
+
+    while((lane = ucp_rkey_get_rma_bw_lane(req->send.rndv_get.rkey, ep,
+                                           &uct_rkey, map)) != UCP_NULL_LANE) {
+        req->send.rndv_get.lane_count++;
+        map |= UCS_BIT(lane);
+    }
+}
+
+static ucp_lane_index_t ucp_rndv_get_next_lane(ucp_request_t *rndv_req, uct_rkey_t *uct_rkey)
+{
+    /* get lane and mask it for next iteration.
+     * next time this lane will not be selected & we continue
+     * with another lane. After all lanes are masked - reset mask
+     * to zero & start from scratch. this way allows to enumerate
+     * all lanes */
+    ucp_ep_h ep = rndv_req->send.ep;
+    ucp_lane_index_t lane;
+
+    lane = ucp_rkey_get_rma_bw_lane(rndv_req->send.rndv_get.rkey, ep,
+                                    uct_rkey, rndv_req->send.rndv_get.lanes_map);
+
+    if (lane == UCP_NULL_LANE) {
+        rndv_req->send.rndv_get.lanes_map = 0;
+        lane = ucp_rkey_get_rma_bw_lane(rndv_req->send.rndv_get.rkey, ep,
+                                        uct_rkey, rndv_req->send.rndv_get.lanes_map);
+    }
+
+    rndv_req->send.rndv_get.lanes_map |= (lane == UCP_NULL_LANE) ? 0 : UCS_BIT(lane);
```

Comment:
how can lane be NULL_LANE at this point?

Reply:
if there are no RMA lanes at all, we should fallback to sending RTR. Where does it happen?

Reply:
It checks for null Lane after call next_lane. We need this evaluation because this call Matt happen after ep became normal (from stub) and we don't know if there are few Lanes

Reply:
so can return "lane" immediately after seeing it's NULL, and let calling function fallback to RTR in https://github.com/openucx/ucx/pull/2116/files#diff-f89c8c0721aa5b2b55347e1612b8395bR298

Reply:
ok, bit refactored code